### PR TITLE
`sqlfunc!` support for SQL funcs that return owned non-Copy values

### PR DESCRIPTION
**Pull request by petrosagg**
### Description

Datums have `From` implementation for primitive, Copy types like `i32`, `f32` etc but it's not possible to provide an implementation of `From<String>` since Datums hold `str` references in their `String` variant.

Traditionally we have been handling this by passing around a `temp_storage: RowArena` parameter that can be thought of as the allocation context of the current function call where a String can be stored in and a reference can be extracted out of it to produce a `Datum::String` variant.

This PR takes this pattern and puts it in a named struct `WithArena<T>` that is simply a wrapper over any T but it also carries with it an allocation context.

This allows us to provide `From<WithArena<String>>` and `From<WithArena<Vec<u8>>`implementations for `Datum` that automatically store their owned data in the arena and construct the appropriate variant.

Using the above machinery an example implementation is provided for `cast_float32_to_string` to keep the PR short. Once this is merged, another PR will follow with all the other String returning functions ported.

### Tips for reviewer

Review commit by commit recommended

### Checklist

- [ ] This PR has adequate test coverage.
- [ ] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: src/expr/src/scalar/func/macros.rs [hunk 1]
```diff
@@ -106,7 +106,7 @@ macro_rules! sqlfunc {
                             .try_into()
                             .expect("expression already typechecked");
 
-                        super::$fn_name(a).map(|r| r.into())
+                        super::$fn_name(a).map(move |r| WithArena::new(temp_storage, r).into())
```
Comment (sploiselle) [inline src/expr/src/scalar/func/macros.rs:109]: As this macro gets more complex, do you think it's possible to keep inline docs around what occurring here? Took me a few min to page some of this back in and am a little worried that the bus factor here is pretty high.

**Review by sploiselle — APPROVED** [review]
Really clean solution! Comments about docs will likely keep coming but are not-blocking. I know this is an evolving piece of code, so don't want to get you to invest too much in explaining things that might change in the near future.

**Comment by petrosagg** [comment]
@sploiselle please keep the push for docs coming! I have been looking at this code for too long and it's not clear to me what warrants explanatory comments or not. I expanded the problematic area in multiple statements and added inline comments explaining what's going on

**Review by sploiselle — COMMENTED** [review]
Perfect, tysm! 👏

**Pull request merged**